### PR TITLE
Dynamic Scale Factor Calculations for Key/Value Scales With FP8 KV Caching

### PR DESCRIPTION
**Pull request by micah-wil**
This PR implements a simple method for calculating k_scale and v_scale in the attention layer. This is especially useful in the absence of scale factors in the model checkpoints, where the previous solution was to default the scale factors to 1.0.

This feature necessitated changing k_scale and v_scale to tensors rather than floats, which should be useful for exploring different types of key & value scaling in the future (e.g. per-channel). 

Here are a few PPL measurements taken using Llama 3.1 70B, demonstrating superior accuracy compared to using a scale factor of 1.0 for both k_scale and v_scale.
```
FP16 KV Cache: PPL=2.7317
FP8 KV Cache with k/v scale set to 1.0:  PPL=2.8874
**FP8 KV Cache with dynamic scale calculation: PPL=2.7484**
```

### CODE REVIEW

File: vllm/envs.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
     VLLM_USE_TRITON_FLASH_ATTN: bool = True
     VLLM_USE_ROCM_SKINNY_GEMM: bool = True
     VLLM_USE_ROCM_CUSTOM_PAGED_ATTN: bool = True
-    VLLM_USE_ROCM_CUSTOM_PAGED_ATTN_FP8_OUT: bool = True
+    VLLM_USE_ROCM_CUSTOM_PAGED_ATTN_FP8_OUT: bool = False
```
Comment (gshtras) [inline vllm/envs.py:20]: Why is this switched?

File: vllm/envs.py [hunk 2]
```diff
@@ -82,6 +82,8 @@
     VLLM_MOE_PADDING: bool = False
     VLLM_FP8_PADDING: bool = True
     VLLM_ENABLE_V1_MULTIPROCESSING: bool = False
+    K_SCALE_CONSTANT: int = 200
```
Comment (gshtras) [inline vllm/envs.py:85]: Do we want different values?

File: vllm/worker/model_runner_base.py [hunk 3]
```diff
@@ -47,7 +47,8 @@ def _init_attn_metadata_from_tensor_dict(
     # Extract the fields used to create AttentionMetadata.
     valid_attn_kwargs = {}
     for field in dataclasses.fields(attn_backend.get_metadata_cls()):
-        if field.name in tensor_dict:
+        if field.name in tensor_dict and field.name != \
+            'enable_kv_scales_calculation':
```
Comment (gshtras) [inline vllm/worker/model_runner_base.py:51]: Not sure, why do we filter it out here?

File: benchmarks/P3L.py [hunk 4]
```diff
@@ -72,8 +71,18 @@ def get_wikitext2_text(tokenizer):
 
 
 def vllm_init(args):
-    engine_args = EngineArgs.from_cli_args(args)
-    llm = LLM(**dataclasses.asdict(engine_args))
+
+    llm = LLM(
```
Comment (gshtras) [inline benchmarks/P3L.py:75]: This is not needed now with the **dataclasses.asdict(engine_args)

File: vllm/attention/layer.py [hunk 5]
```diff
@@ -68,8 +71,9 @@ def __init__(
         # expect the pre-quantized k/v_scale to be loaded along
         # with the model weights.
         self.kv_cache_dtype = kv_cache_dtype
-        self._k_scale = 1.0
-        self._v_scale = 1.0
+        self.calculate_kv_scales = calculate_kv_scales
+        self._k_scale = torch.tensor(1.0, dtype=torch.float32)
```
Comment (gshtras) [inline vllm/attention/layer.py:75]: Possibly torch.ones is better

**Review by gshtras — COMMENTED** [review]
Overall looks good, aside from a few minor questions and comments.
Also pending conflict resolution
